codegen/project_generator.go

```diff
@@ -49,90 +59,102 @@ func NewProjectGenerator(srcPath, projectName string, opts ...ProjectGeneratorOp
 
 	return pj
 }
-
 func (g *ProjectGenerator) GenerateProject() error {
+	engine := &Engine{Manifest: &manifest.Manifest{}}
+	utils["getEngine"] = engine.GetEngine
+	directories := []string{
+		".cargo",
+		"proto",
+		"src",
+		filepath.Join("src", "abi"),
+		filepath.Join("src", "pb"),
+	}
 	if _, err := os.Stat(g.srcPath); errors.Is(err, os.ErrNotExist) {
 		fmt.Printf("Creating missing %q folder\n", g.srcPath)
 		if err := os.MkdirAll(g.srcPath, os.ModePerm); err != nil {
 			return fmt.Errorf("creating src directory %v: %w", g.srcPath, err)
 		}
 	}
 
-	fullPath := filepath.Join(g.srcPath, g.ProjectName)
-	if _, err := os.Stat(fullPath); errors.Is(err, os.ErrNotExist) {
+	projectPath := filepath.Join(g.srcPath, g.ProjectName)
+	if _, err := os.Stat(projectPath); errors.Is(err, os.ErrNotExist) {
 		fmt.Printf("Creating missing %q folder\n", g.srcPath)
-		if err := os.MkdirAll(fullPath, 0755); err != nil {
-			return fmt.Errorf("creating missing %q folder: %w", g.srcPath, err)
+		if err := os.MkdirAll(projectPath, 0755); err != nil {
+			return fmt.Errorf("creating missing %q folder: %w", projectPath, err)
 		}
 	}
 
-	fmt.Printf("Generating files in %q\n", fullPath)
-
-	srcDir := filepath.Join(fullPath, "src")
-	if _, err := os.Stat(srcDir); errors.Is(err, os.ErrNotExist) {
-		fmt.Printf("Creating missing %q folder\n", srcDir)
-		if err := os.MkdirAll(srcDir, os.ModePerm); err != nil {
-			return fmt.Errorf("creating src directory %v: %w", srcDir, err)
-		}
-	} else {
-		fmt.Println("src directory already exists, skipping")
+	// generate template from ./templates
+	templateFiles, err := template.New("templates").Funcs(utils).ParseFS(templates, "*/*.gotmpl", "*/*/*.gotmpl", "*/*/*/*.gotmpl")
+	if err != nil {
+		return fmt.Errorf("instantiate template: %w", err)
 	}
 
-	protoDir := filepath.Join(fullPath, "proto")
-	if _, err := os.Stat(protoDir); errors.Is(err, os.ErrNotExist) {
-		fmt.Printf("Creating missing %q folder\n", protoDir)
-		if err := os.MkdirAll(protoDir, os.ModePerm); err != nil {
-			return fmt.Errorf("creating proto directory %v: %w", protoDir, err)
+	// create directories
+	for _, dir := range directories {
+		dirPath := filepath.Join(projectPath, dir)
+		if _, err := os.Stat(dirPath); errors.Is(err, os.ErrNotExist) {
+			fmt.Printf("Creating missing %q folder\n", dirPath)
+			if err := os.MkdirAll(dirPath, os.ModePerm); err != nil {
+				return fmt.Errorf("creating directory %v: %w", dirPath, err)
```

Comment:
```suggestion
				return fmt.Errorf("creating directory %q: %w", dirPath, err)
```

`%q` is like `%s` but will add double quotes which make it easier to spot empty string and make the "argument" is easier to spot.